src/Materials/customShaderCode.ts

```diff
@@ -0,0 +1,295 @@
+            //     this.CustomParts.Fragment_Custom_Alpha = shaderPart.replace("result", "alpha");
+            //     return this;
+            // }
+
+            // public Fragment_Before_FragColor(shaderPart: string): CustomMaterial {
+            //     this.CustomParts.Fragment_Before_FragColor = shaderPart.replace("result", "color");
+            //     return this;
+            // }
+
+            // public Vertex_Before_PositionUpdated(shaderPart: string): CustomMaterial {
+            //     this.CustomParts.Vertex_Before_PositionUpdated = shaderPart.replace("result", "positionUpdated");
+            //     return this;
+            // }
+
+            // public Vertex_Before_NormalUpdated(shaderPart: string): CustomMaterial {
+            //     this.CustomParts.Vertex_Before_NormalUpdated = shaderPart.replace("result", "normalUpdated");
+            //     return this;
+            // }
+        }
+
+        // add uniforms
+        pieces[ShaderCodePoint.Vertex_Definitions] = (this._customUniform ? this._customUniform.join("\n") : "") +
+            pieces[ShaderCodePoint.Vertex_Definitions];
+        pieces[ShaderCodePoint.Fragment_Definitions] = (this._customUniform ? this._customUniform.join("\n") : "") +
+            pieces[ShaderCodePoint.Fragment_Definitions];
+
+        let VertexShader = this.VertexShader;
+        let FragmentShader = this.FragmentShader;
+
+        for (const point of ShaderCodePointValues) {
+            if (isNaN(Number(point))) {
+                continue;
+            }
+            const piece = pieces[point as ShaderCodePoint];
+            if (!piece) {
+                continue;
+            }
+            if (point.indexOf('Vertex_') !== -1) {
+                VertexShader = VertexShader.replace(
+                    '#define ' + point,
+                    piece
+                );
+            }
+            else if (point.indexOf('Fragment_') !== -1) {
+                FragmentShader = FragmentShader.replace(
+                    '#define ' + ShaderCodePoint[point as ShaderCodePointKeys],
+                    piece
+                );
+            }
+        }
+
+        Effect.ShadersStore[name + "VertexShader"] = VertexShader;
+        Effect.ShadersStore[name + "PixelShader"] = FragmentShader;
+
+        // TODO: why are these special cases?
+        // if (this.CustomParts.Vertex_After_WorldPosComputed) {
+        //     Effect.ShadersStore[name + "VertexShader"] = Effect.ShadersStore[name + "VertexShader"].replace('#define CUSTOM_VERTEX_UPDATE_WORLDPOS', this.CustomParts.Vertex_After_WorldPosComputed);
+        // }
+        // if (this.CustomParts.Fragment_Before_Fog) {
+        //     Effect.ShadersStore[name + "PixelShader"] = Effect.ShadersStore[name + "PixelShader"].replace('#define CUSTOM_FRAGMENT_BEFORE_FOG', this.CustomParts.Fragment_Before_Fog);
+        // }
+
+        this._isCreatedShader = true;
+        this._createdShaderName = name;
+
+        return name;

```

Comment:
We need a way to track the generated code better so we can remove it if required. Why not returning a fully functional ShaderMaterial?